File: src/services/coordinator.ts
/**
 * 核心协调器服务
 */

import { invoke } from '@tauri-apps/api/core'

import type { BlackholeBreakerConfig } from '@/services/blackhole-breaker'
import type { ResolvedEgressIdentity } from '@/services/egress-identity'
import {
  normalizeIpReputationConfig,
  serializeIpReputationConfig,
  type IpReputationConfig,
} from '@/services/ip-reputation'
import type { LocalStealthConfig } from '@/services/local-stealth'
import type { MultipathConfig } from '@/services/multipath'
import type { BindingInfo, SessionAffinityConfig } from '@/services/session-affinity'
import type { TimezoneSpoofConfig } from '@/services/timezone-spoof'

export type {
  MultipathConfig,
  NodePool,
  PathNode,
  PoolType,
  SessionBinding,
  SlicingStrategy,
} from '@/services/multipath'
export type {
  ConnectionBindingConfig,
  DomainBindingRule,
  ProcessBindingRule,
  SessionAffinityConfig,
} from '@/services/session-affinity'

/**
 * 协调器状态
 */
export interface CoordinatorResolvedEgressIdentity extends ResolvedEgressIdentity {
  sourceGroupName?: string | null
  sourceGroupSelectedNode?: string | null
}

export interface CoordinatorBindingInfo extends BindingInfo {
  sourceGroupName?: string | null
  sourceGroupSelectedNode?: string | null
}

export interface StableEgressBackwriteStatus {
  domainPatternAssignments: CoordinatorResolvedEgressIdentity[]
  domainRuleBindings: CoordinatorBindingInfo[]
}

export interface CoordinatorRuntimeState {
  egressIdentityAssignments: CoordinatorResolvedEgressIdentity[]
  sessionAffinityBindings: CoordinatorBindingInfo[]
  stableEgressBackwrite: StableEgressBackwriteStatus
}

export interface CoordinatorStatus {
  initialized: boolean
  securityEnabled: boolean
  securityCompromised: boolean
  antiProbeEnabled: boolean
  tlsFingerprint: string | null
  egressIdentityEnabled: boolean
  sessionAffinityEnabled: boolean
  egressIdentityActiveAssignments: number
  sessionAffinityActiveBindings: number
  runtimeState: CoordinatorRuntimeState
  multipathEnabled: boolean
  trafficObfuscationEnabled: boolean
  xdpEnabled?: boolean
  xdpRunning?: boolean
}

/**
 * 高级配置
 */
export interface AdvancedConfig {
  security: SecurityConfig
  multipath: MultipathConfig
  session_affinity: SessionAffinityConfig
  egress_identity: EgressIdentityConfig
  egress_monitor: EgressMonitorConfig
  dns: AdvancedDnsConfig
  traffic_obfuscation: TrafficObfuscationConfig
  traffic_padding: TrafficPaddingConfig
  security_policies: ISecurityPolicy[]
  residential_pool: ResidentialProxyPool
  ip_reputation: IpReputationConfig
  blackhole_breaker: BlackholeBreakerConfig
  timezone_spoof: TimezoneSpoofConfig
  local_stealth: LocalStealthConfig
  ingress_countermeasure: IngressCountermeasureConfig
  xdp?: XdpConfig
}

export interface IngressCountermeasureConfig {
  enabled: boolean
  classifierThresholds: ClassifierThresholds
  personaProfiles: PersonaProfile[]
  deceptionMode: DeceptionMode
  responseDelayRanges: ResponseDelayRanges
  fakeSurfacePolicies: FakeSurfacePolicy[]
  egressStabilitySupport: EgressStabilitySupportConfig
}

export interface ClassifierThresholds {
  lowConfidence: number
  mediumConfidence: number
  highConfidence: number
}

export interface PersonaProfile {
  id: string
  label: string
  tone: PersonaTone
  surfaceBias: SurfaceBias
}

export type PersonaTone = 'restrained' | 'neutral' | 'helpful'

export type SurfaceBias = 'decoy' | 'balanced' | 'production'

export type DeceptionMode =
  | 'disabled'
  | 'observeOnly'
  | 'decoyPreferred'
  | 'decoyOnly'

export interface ResponseDelayRanges {
  softDelayMinMs: number
  softDelayMaxMs: number
  hardDelayMinMs: number
  hardDelayMaxMs: number
}

export interface FakeSurfacePolicy {
  surface: string
  priority: number
  enabled: boolean
}

export interface EgressStabilitySupportConfig {
  enabled: boolean
  rebindGracePeriodMs: number
  connectionWarmupMs: number
}

export interface SecurityConfig {
  enabled: boolean
  anti_probe: AntiProbeConfig
  tls_fingerprint: string | null
  config_decoy: ConfigDecoyConfig
  sniffer: SnifferConfig
  obfuscation: ObfuscationConfig
}

export interface SnifferConfig {
  enabled: boolean
  overrideDest: boolean
  forceDomain: string[]
  skipDomain: string[]
  parsePureIp: boolean
  forceDnsMapping: boolean
  sniffing: string[]
}

export interface AntiProbeConfig {
  enabled: boolean
  secret_key: string
  time_window: number
  whitelist: string[]
  strict_mode: boolean
}

export interface ConfigDecoyConfig {
  enabled: boolean
  decoy_path: string | null
}

export type ResidentialProxyType = 'socks5' | 'http' | 'ss' | 'vmess' | 'trojan'

export interface ResidentialProxy {
  name: string
  proxyType: ResidentialProxyType
  server: string
  port: number
  username?: string
  password?: string
  cipher?: string
  uuid?: string
  trojanPassword?: string
  tls?: boolean
  sni?: string
  skipCertVerify?: boolean
  region?: string
  enabled: boolean
}

export interface ResidentialProxyPool {
  enabled: boolean
  proxies: ResidentialProxy[]
}

export type ObfuscationLevel = 'none' | 'low' | 'medium' | 'high' | 'paranoid'

export interface ObfuscationConfig {
  enabled: boolean
  level: ObfuscationLevel
  autoAdjust: boolean
}

export type TrafficObfuscationProfile =
  | 'none'
  | 'conservative'
  | 'aggressive'
  | 'custom'

export interface TrafficObfuscationConfig {
  enabled: boolean
  profile: TrafficObfuscationProfile
  padding: TrafficPaddingConfig
  timing: TimingJitterConfig
  direction: DirectionObfuscationConfig
}

export type PaddingIntensity =
  | 'Low'
  | 'Medium'
  | 'High'
  | { Custom: number }

export type FrequencyType = 'Time' | 'Request' | 'Random'

export interface PaddingFrequency {
  freqType: FrequencyType
  interval: number
}

export type PaddingTiming = 'Before' | 'After' | 'Random'

export interface PerformanceControl {
  maxBandwidth: number
  maxCpuUsage: number
  maxMemory: number
  autoDowngrade: boolean
}

export interface TrafficPaddingConfig {
  enabled: boolean
  minSize: number
  maxSize: number
  encrypt: boolean
  intensity: PaddingIntensity
  frequency: PaddingFrequency
  timing: PaddingTiming
  smartPadding: boolean
  performanceControl: PerformanceControl
}

export type JitterMode = 'uniform' | 'gaussian' | 'pareto'

export interface TimingJitterConfig {
  enabled: boolean
  mode: JitterMode
  minDelayMs: number
  maxDelayMs: number
  batchWindowMs: number
}

export type DirectionMode = 'mirror' | 'pad' | 'random'

export interface DirectionObfuscationConfig {
  enabled: boolean
  mode: DirectionMode
  mirrorRatio: number
  padToSize: number
}

export type IpType = 'Datacenter' | 'Residential' | 'Mobile' | 'Unknown'

export interface AdvancedDnsConfig {
  enable_cache: boolean
  enable_prefetch: boolean
  enable_health_check: boolean
  prefetch_interval: number
  health_check_interval: number
  routing_mode: DnsRoutingMode
  leak_protection_level: DnsLeakProtectionLevel
}

export type DnsRoutingMode = 'speed' | 'privacy' | 'balanced' | 'custom'

export type DnsLeakProtectionLevel = 'none' | 'basic' | 'strict' | 'paranoid'

export interface EgressIdentityConfig {
  enabled: boolean
  default_profile: string | null
  profiles: EgressIdentityProfile[]
  app_rules: AppEgressRule[]
  shortcut_rules: ShortcutEgressRule[]
}

export interface EgressIdentityProfile {
  id: string
  name: string
  enabled: boolean
  preferred_nodes: string[]
  preferred_pools: string[]
  required_ip_type: IpType | null
  max_fraud_score: number | null
  dns_policy: DnsPolicy
  tls_fingerprint: string | null
  session_policy: IdentitySessionPolicy
  failover_policy: EgressFailoverPolicy
  allowed_nodes?: string[]
  strict_node_scope?: boolean
  use_residential_chain?: boolean
  residential_proxy_name?: string | null
  description: string
}

export interface AppEgressRule {
  process_name: string | null
  exe_path: string | null
  domains: string[]
  profile_id: string
  priority: number
  enabled: boolean
}

export interface ShortcutEgressRule {
  shortcut_id: string
  profile_id: string
  enabled: boolean
}

export interface DnsPolicy {
  mode: DnsMode
  force_remote_dns: boolean
}

export type DnsMode = 'Inherit' | 'Hijack' | 'Remote'

export interface IdentitySessionPolicy {
  strict_affinity: boolean
  ttl_override: number | null
}

export type EgressFailoverPolicy = 'Block' | 'Manual' | 'AutoSwitch'

export type RebindStrategyType = 'smart' | 'round-robin'

export interface EgressMonitorConfig {
  enabled: boolean
  probeIntervalSecs: number
  autoRebindOnChange: boolean
  notifyOnChange: boolean
  probeTimeoutSecs: number
  watchPollIntervalSecs: number
  watchDebounceSecs: number
  rebindStrategy: RebindStrategyType
}

export interface EgressIpProbeResult {
  ip: string
  countryCode: string | null
  probedAtMs: number
  latencyMs: number
}

export interface EgressIpChangeEvent {
  previousIp: string
  currentIp: string
  previousCountry: string | null
  currentCountry: string | null
  timestampMs: number
  autoRebindApplied: boolean
}

export interface EgressMonitorStats {
  totalProbes: number
  successfulProbes: number
  failedProbes: number
  ipChangeCount: number
  autoRebindCount: number
  lastProbe: EgressIpProbeResult | null
  lastChange: EgressIpChangeEvent | null
  uptimeSecs: number
}

export interface XdpConfig {
  enabled: boolean
  interface: string
  mode: XdpMode
  queue_size: number
}

function normalizeAdvancedConfig(config: AdvancedConfig): AdvancedConfig {
  return {
    ...config,
    ip_reputation: normalizeIpReputationConfig(config.ip_reputation),
  }
}

function serializeAdvancedConfig(config: AdvancedConfig) {
  return {
    ...config,
    ip_reputation: serializeIpReputationConfig(config.ip_reputation),
  }
}

export type XdpMode = 'Native' | 'Skb' | 'Generic'

/**
 * 初始化协调器
 */
export async function coordinatorInitialize(): Promise<void> {
  await invoke('coordinator_initialize')
}

/**
 * 关闭协调器
 */
export async function coordinatorShutdown(): Promise<void> {
  await invoke('coordinator_shutdown')
}

/**
 * 获取高级配置
 */
export async function getAdvancedConfig(): Promise<AdvancedConfig> {
  const config = await invoke<AdvancedConfig>('get_advanced_config')
  return normalizeAdvancedConfig(config)
}

/**
 * 保存高级配置
 */
export async function saveAdvancedConfig(config: AdvancedConfig): Promise<void> {
  await invoke('save_advanced_config', { config: serializeAdvancedConfig(config) })
}

/**
 * 获取推荐配置
 */
export async function getRecommendedAdvancedConfig(): Promise<AdvancedConfig> {
  const config = await invoke<AdvancedConfig>('get_recommended_advanced_config')
  return normalizeAdvancedConfig(config)
}

/**
 * 验证高级配置
 */
export async function validateAdvancedConfig(config: AdvancedConfig): Promise<void> {
  await invoke('validate_advanced_config', { config: serializeAdvancedConfig(config) })
}

/**
 * 获取协调器状态
 */
export async function coordinatorGetStatus(): Promise<CoordinatorStatus> {
  return await invoke('coordinator_get_status')
}

File: src-tauri/src/feat/coordinator.rs
use crate::config::AdvancedConfig;
use crate::core::coordinator_status::CoordinatorStatus;
use crate::core::stable_egress::project_runtime_status;
use crate::core::{CoreManager, coordinator::CoreCoordinator};
use crate::traffic::TrafficObfuscationConfig;
use anyhow::Result;
use once_cell::sync::Lazy;
use std::sync::Arc;

/// 全局协调器实例
static COORDINATOR: Lazy<Arc<CoreCoordinator>> = Lazy::new(|| Arc::new(CoreCoordinator::new()));

/// 获取协调器实例
pub fn get_coordinator() -> Arc<CoreCoordinator> {
    COORDINATOR.clone()
}

/// 从磁盘重新加载 AdvancedConfig 并同步到 coordinator 内存
pub fn sync_coordinator_from_advanced_config() -> Result<()> {
    let path = AdvancedConfig::default_path()?;
    let config = AdvancedConfig::load(&path)?;
    COORDINATOR.hydrate_from_advanced_config(&config)
}

pub async fn sync_coordinator_from_advanced_config_async() -> Result<()> {
    let path = AdvancedConfig::default_path()?;
    let config = AdvancedConfig::load(&path)?;
    COORDINATOR.hydrate_from_advanced_config(&config)?;
    crate::feat::apply_egress_monitor_config(config.egress_monitor.clone()).await?;
    crate::feat::apply_ip_reputation_config(config.ip_reputation.clone()).await?;
    crate::feat::apply_blackhole_breaker_config(config.blackhole_breaker.clone()).await;
    crate::core::security_runtime::apply_local_stealth_config(config.local_stealth.clone()).await;
    let traffic_obfuscation = if config.traffic_obfuscation.enabled {
        config.traffic_obfuscation.clone()
    } else if config.traffic_padding.enabled {
        TrafficObfuscationConfig::from_legacy_padding(&config.traffic_padding)
    } else {
        config.traffic_obfuscation.clone()
    };
    crate::feat::apply_traffic_obfuscation_config(traffic_obfuscation).await?;
    crate::feat::get_session_affinity_manager()
        .update_config(config.session_affinity)
        .await?;
    Ok(())
}

/// 保存高级配置（业务逻辑）
pub async fn save_advanced_config(config: &AdvancedConfig) -> Result<()> {
    config.validate()?;

    let path = AdvancedConfig::default_path()?;
    config.save(&path)?;

    crate::feat::save_dns_config_mapping(&config.dns.to_dns_config_mapping()).await?;
    crate::feat::apply_egress_monitor_config(config.egress_monitor.clone()).await?;
    crate::feat::apply_ip_reputation_config(config.ip_reputation.clone()).await?;
    crate::feat::apply_blackhole_breaker_config(config.blackhole_breaker.clone()).await;
    crate::core::security_runtime::apply_local_stealth_config(config.local_stealth.clone()).await;
    #[cfg(target_os = "linux")]
    get_coordinator().xdp_manager().update_config(config.xdp.clone());

    // 同步流量混淆配置：优先使用 traffic_obfuscation，兼容旧 traffic_padding
    let obf_config = if config.traffic_obfuscation.enabled {
        config.traffic_obfuscation.clone()
    } else if config.traffic_padding.enabled {
        TrafficObfuscationConfig::from_legacy_padding(&config.traffic_padding)
    } else {
        config.traffic_obfuscation.clone()
    };
    crate::feat::apply_traffic_obfuscation_config(obf_config).await?;

    // 应用到协调器
    COORDINATOR.apply_advanced_config(config)?;

    crate::feat::get_session_affinity_manager()
        .update_config(config.session_affinity.clone())
        .await?;

    CoreManager::global().update_config_checked().await?;

    Ok(())
}

/// 获取协调器状态（业务逻辑）
pub async fn coordinator_get_status() -> Result<CoordinatorStatus> {
    let _ = sync_coordinator_from_advanced_config_async().await;
    let config = COORDINATOR.get_advanced_config();
    let security_compromised = crate::security::is_security_compromised();
    let runtime_state = project_runtime_status(
        COORDINATOR.egress_identity_manager().get_active_assignments(),
        crate::feat::get_session_affinity_manager().get_all_bindings().await?,
    )
    .await;
    let egress_identity_active_assignments = runtime_state.egress_identity_assignments.len();
    let session_affinity_active_bindings = runtime_state.session_affinity_bindings.len();

    Ok(CoordinatorStatus {
        initialized: true,
        security_enabled: config.security.enabled,
        security_compromised,
        anti_probe_enabled: config.security.anti_probe.enabled,
        tls_fingerprint: config.security.tls_fingerprint.clone(),
        egress_identity_enabled: config.egress_identity.enabled,
        session_affinity_enabled: config.session_affinity.enabled,
        egress_identity_active_assignments,
        session_affinity_active_bindings,
        runtime_state,
        multipath_enabled: config.multipath.enabled,
        traffic_obfuscation_enabled: config.traffic_obfuscation.enabled,
        honeypot_enabled: config.security.honeypot.enabled,
        self_destruct_enabled: config.security.self_destruct.enabled,
        #[cfg(target_os = "linux")]
        xdp_enabled: config.xdp.enabled,
        #[cfg(target_os = "linux")]
        xdp_running: COORDINATOR.xdp_manager().is_running(),
    })
}
